### PR TITLE
Add Microsoft.Build.NuGetSdkResolver project.

### PR DESCRIPTION
**Pull request by AndyGerlicher**
https://github.com/NuGet/Home/issues/6799

Still work in progress. Haven't added tests yet.

### CODE REVIEW

**Comment by rohit21agrawal** [comment]
running through CI to see if artifacts and nupkgs are generated correctly

**Comment by AndyGerlicher** [comment]
@rohit21agrawal does this look like something my change broke?

### Error message
`Expected False because newtonsoft.json should NOT be in TestProject, but found True.`
### Stack trace
```
at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message)
at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
at FluentAssertions.Primitives.BooleanAssertions.BeFalse(String because, Object[] becauseArgs)
at NuGet.Tests.Apex.Utils.AssetPackageNotInPackagesConfig(VisualStudioHost visualStudio, ProjectTestExtension project, String packageName, ILogger logger) in E:\A\_work\2\s\test\NuGet.Tests.Apex\NuGet.Tests.Apex\NuGetIVsApiTests\Utils.cs:line 196
at NuGet.Tests.Apex.NuGetUITestCase.UninstallPackageFromUI() in E:\A\_work\2\s\test\NuGet.Tests.Apex\NuGet.Tests.Apex\NuGetIVsApiTests\NuGetUITestCase.cs:line 111
```

**Comment by rohit21agrawal** [comment]
no this test is flaky. so far everything seems to be fine.

**Comment by rohit21agrawal** [comment]
@AndyGerlicher  this is the error :

``` NuGetSdkLogger_Tests.cs(139,39): error CS0122: 'NuGetSdkLogger' is inaccessible due to its protection level [E:\A\_work\9\s\test\NuGet.Core.Tests\Microsoft.Build.NuGetSdkResolver.Tests\Microsoft.Build.NuGetSdkResolver.Test.csproj] ```

**Comment by AndyGerlicher** [comment]
Looks like I have my InternalsVisisbleTo wrong. What's the PK of the test projects? Looking at other examples I see at least two?
```
0024000004800000940000000602000000240000525341310004000001000100c547cac37abd99c8db225ef2f6c8a3602f3b3606cc9891605d02baa56104f4cfc0734aa39b93bf7852f7d9266654753cc297e7d2edfe0bac1cdcf9f717241550e0a7b191195b7667bb4f64bcb8e2121380fd1d9d46ad2d92d2d15605093924cceaf74c4861eff62abf69b9291ed0a340e113be11e6a7d3113e92484cf7045cc7
002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293
```

**Comment by rohit21agrawal** [comment]
@AndyGerlicher  it's

``` 
0024000004800000940000000602000000240000525341310004000001000100b5fc90e7027f67871e773a8fde8938c81dd402ba65b9201d60593e96c492651e889cc13f1415ebb53fac1131ae0bd333c5ee6021672d9718ea31a8aebd0da0072f25d87dba6fc90ffd598ed4da35e44c398c454307e8e33b8426143daec9f596836f97c8f74750e5975c64e2189f45def46b2a2b1247adc3652bf5c308055da9
```

**Comment by AndyGerlicher** [comment]
Thanks! Need CI again though please.

**Comment by rohit21agrawal** [comment]
@AndyGerlicher are you going to resume working on this?

**Comment by AndyGerlicher** [comment]
@rohit21agrawal sorry for the delay was out last week. Can you CI approve this again please? Was able to get the tests running locally and now they pass.

**Comment by rohit21agrawal** [comment]
@AndyGerlicher the build on linux is still failing 

You can run it locally:

```
scripts/funcTests/runFuncTests.sh
```

**Comment by AndyGerlicher** [comment]
@rohit21agrawal is there a way to get the build logs? When I click on the "Tests on Linux" Details all I see is that 420 passed 80 not executed. I don't see any indication of an error or failure?

**Comment by rohit21agrawal** [comment]
https://devdiv.visualstudio.com/DevDiv/_build?buildId=1629905

These are the build logs. In the Phase Tests on Linux, look at the logs for the step Run tests

There are two sets of tests that run on Linux and generate two different results file. The first set didn’t even get to running any tests, and the second one completed successfully, that’s why you don’t see failures in tests.

**Comment by AndyGerlicher** [comment]
@rohit21agrawal ok how about now?

**Comment by rohit21agrawal** [comment]
@AndyGerlicher tests look good, whenever you are done with code changes, remove the WIP title from PR , and we'll start the code review.

**Comment by AndyGerlicher** [comment]
This is good to review. I just verified the end to end scenario (this change + my MSBuild resolver manifest change).

Do I need to update a swr file or something to get it to appear in the vsix? And someone needs to drop a an xml file into `MSBuild\15.0\Bin\SdkResolvers\...` Makes sense for you guys to do it since the feature wouldn't work without this resolver and you control the path to NuGet. But I'm fine either way really.

**Comment by AndyGerlicher** [comment]
Also @jeffkl for review

File: src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            fullPath = GetPathOfFileAbove(file, startingDirectory);
+
+            return fullPath != string.Empty;
+        }
+
+        /// <summary>
+        /// Locate a file in either the directory specified or a location in the
+        /// directory structure above that directory.
+        /// </summary>
+        private static string GetDirectoryNameOfFileAbove(string startingDirectory, string fileName)
```
Comment (jeffkl) [inline src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs:101]: Is there a NuGet API that can do this?  I know they have code to search for NuGet.config

**Review by jeffkl — APPROVED** [review]

**Pull request closed without merging**